sky/provision/aws/instance.py

```diff
@@ -184,20 +178,125 @@ def cleanup_ports(
+    # instance's security group, so here we move to it.
+    sg = _get_sg_from_name(ec2, expected_sg_name)
+    instance.modify_attribute(Groups=[sg.id])
+    return sg
+
+
+def open_ports(
+    cluster_name_on_cloud: str,
+    ports: List[str],
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> None:
+    """See sky/provision/__init__.py"""
+    assert provider_config is not None, (cluster_name_on_cloud, provider_config)
+    region = provider_config['region']
+    ec2 = _default_ec2_resource(region)
+    sg_name = provider_config['security_group']['GroupName']
+    filters = [
+        {
+            'Name': 'instance-state-name',
+            # exclude 'shutting-down' or 'terminated' states
+            'Values': ['pending', 'running', 'stopping', 'stopped'],
+        },
+        *_cluster_name_filter(cluster_name_on_cloud),
+    ]
+    instances = _filter_instances(ec2, filters, None, None)
+    if len(list(instances)) != 1:
+        logger.warning(
+            f'Expected 1 instance with cluster name {cluster_name_on_cloud}, '
+            f'but found {len(list(instances))}. Skip creating security group.')
+        return None
+    instance = list(instances)[0]
+    sg = _maybe_move_to_new_sg(ec2, instance, sg_name)
+    if sg is None:
+        logger.warning('Find new security group failed. Skip open ports.')
+        return
+
+    existing_ports: Set[int] = set()
+    for existing_rule in sg.ip_permissions:
+        # Skip any non-tcp rules.
+        if existing_rule['IpProtocol'] != 'tcp':
+            continue
+        # Skip any rules that don't have a FromPort or ToPort.
+        if 'FromPort' not in existing_rule or 'ToPort' not in existing_rule:
+            continue
+        existing_ports.update(
+            range(existing_rule['FromPort'], existing_rule['ToPort'] + 1))
+    ports_to_open = resources_utils.parse_port_set(
+        resources_utils.parse_ports(ports) - existing_ports)
+
+    ip_permissions = []
+    for port in ports_to_open:
+        if port.isdigit():
+            from_port = to_port = port
+        else:
+            from_port, to_port = port.split('-')
+        ip_permissions.append({
+            'FromPort': int(from_port),
+            'ToPort': int(to_port),
+            'IpProtocol': 'tcp',
+            'IpRanges': [{
+                'CidrIp': '0.0.0.0/0'
+            }],
+        })
+
+    # For the case when every new ports is already opened.
+    if ip_permissions:
+        sg.authorize_ingress(IpPermissions=ip_permissions)
+
+
+def cleanup_ports(
+    cluster_name_on_cloud: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> None:
+    """See sky/provision/__init__.py"""
+    assert provider_config is not None, cluster_name_on_cloud
+    region = provider_config['region']
+    ec2 = _default_ec2_resource(region)
+    sg_name = provider_config['security_group']['GroupName']
+    if sg_name == backend_utils.DEFAULT_AWS_SG_NAME:
```

Comment:
We don't want to import `backend_utils` in the provisioner, as the provisioner is the lower-level API. We can place the constant in some `constants.py`

Reply:
I cannot find a suitable place for it... `sky/skylet/constants.py` is not valid since we require `common_utils.py` to generate the name, which will lead to a circular import error; other `constants.py` seems a little bit strange for this constant. Currently, I get rid of it by using another method that doesn't require this constant.
<img width="271" alt="image" src="https://github.com/skypilot-org/skypilot/assets/74357442/e8e7c446-d7c1-4c51-9e2d-e60f1195f1e2">